{{cookiecutter.repostory_name}}/app/src/{{cookiecutter.django_project_name}}/{{cookiecutter.django_default_app_name}}/constance_prometheus.py

```diff
@@ -0,0 +1,133 @@
+"""
+Export django-constance variables as Prometheus metrics.
+
+By default this module exports all constance variables which are of the ``int``,
+``float``, ``bool`` or ``decimal.Decimal`` types and not included in the
+``CONSTANCE_PROMETHEUS_BLACKLIST`` sequence.
+
+The variables will appear named ``constance_config_VARIABLE_KEY`` and represented
+as Gauge metrics.
+
+The ``CONSTANCE_PROMETHEUS_WHITELIST`` sequence may specify additional variables
+to be exported. Their values will be cast into ``str`` and appended to the metric
+name (e.g. ``VAR="foo bar"`` will become ``constance_config_VAR_foo_bar``) and
+get 1.0 value. Once changed, the old metric becomes 0.0 and new one will appear,
+allowing to present different values on a graph.
+
+"""
+import os
+import re
+
+from constance import config, settings as constance_settings
+from constance.signals import config_updated
+from decimal import Decimal
+from django.conf import settings
+from django.dispatch import receiver
+from prometheus_client import REGISTRY, CollectorRegistry, Gauge
+from typing import Any
+
+from . import metrics
+
+
+def is_blacklisted(key: str) -> bool:
+    return key in getattr(settings, "CONSTANCE_PROMETHEUS_BLACKLIST", [])
+
+
+def is_whitelisted(key: str) -> bool:
+    return key in getattr(settings, "CONSTANCE_PROMETHEUS_WHITELIST", [])
+
+
+def is_of_automatically_monitored_type(value: Any) -> bool:
+    return isinstance(value, (int, float, bool, Decimal))
+
+
+def gets_monitored(key: str, value: Any) -> bool:
+    return not is_blacklisted(key) and (
+        is_of_automatically_monitored_type(value) or is_whitelisted(key)
+    )
+
+
+METRIC_NAME_FILTER = re.compile(r"\W", re.ASCII)
+
+
+class Metric:
+    PREFIX = "constance_config"
+
+    def __init__(self, config_key: str):
+        self.config_key = config_key
+        self.name = self._get_name()
+        self.description = constance_settings.CONFIG[self.config_key][1]
+        value = getattr(config, self.config_key)
+        if is_of_automatically_monitored_type(value):
+            self._metric = Gauge(self.name, self.description)
+            self.store = self._metric.set
+        else:
+            # a str or custom type; cast everything to str
+            value = str(value)
+            self.name = self._get_name(value)
+            self._metric = Gauge(self.name, self.description)
```

Comment:
I'm worried about the `multiprocess_mode` being let default here, which is `all`. I think what we need is `livemax`. The current behaviour will be:

'all': Default. Return a timeseries per process (alive or dead), labelled by the process's pid (the label is added internally).

If constance yields different values in different processes, it's either a misconfiguration or a bug, not sure we should be detecting it here

Reply:
It doesn't help there, either. Adding the mode, I still get a duplicate error when changing the variable back to previously used value:

```
Environment:


Request Method: POST
Request URL: http://localhost:8000/admin/constance/config/

Django Version: 3.2.12
Python Version: 3.10.4
Installed Applications:
['django_prometheus',
 'django.contrib.admin',
 'django.contrib.auth',
 'django.contrib.contenttypes',
 'django.contrib.sessions',
 'django.contrib.messages',
 'django.contrib.staticfiles',
 'django_extensions',
 'django_probes',
 'constance',
 'project.core',
 'corsheaders']
Installed Middleware:
['corsheaders.middleware.CorsMiddleware',
 'django_prometheus.middleware.PrometheusBeforeMiddleware',
 'django.middleware.security.SecurityMiddleware',
 'django.contrib.sessions.middleware.SessionMiddleware',
 'django.middleware.common.CommonMiddleware',
 'django.middleware.csrf.CsrfViewMiddleware',
 'django.contrib.auth.middleware.AuthenticationMiddleware',
 'django.contrib.messages.middleware.MessageMiddleware',
 'django.middleware.clickjacking.XFrameOptionsMiddleware',
 'django_prometheus.middleware.PrometheusAfterMiddleware']



Traceback (most recent call last):
  File "/home/emes/venvs/reef/lib/python3.10/site-packages/django/core/handlers/exception.py", line 47, in inner
    response = get_response(request)
  File "/home/emes/venvs/reef/lib/python3.10/site-packages/django/core/handlers/base.py", line 181, in _get_response
    response = wrapped_callback(request, *callback_args, **callback_kwargs)
  File "/home/emes/venvs/reef/lib/python3.10/site-packages/django/utils/decorators.py", line 130, in _wrapped_view
    response = view_func(request, *args, **kwargs)
  File "/home/emes/venvs/reef/lib/python3.10/site-packages/django/views/decorators/cache.py", line 44, in _wrapped_view_func
    response = view_func(request, *args, **kwargs)
  File "/home/emes/venvs/reef/lib/python3.10/site-packages/django/contrib/admin/sites.py", line 232, in inner
    return view(request, *args, **kwargs)
  File "/home/emes/venvs/reef/lib/python3.10/site-packages/django/utils/decorators.py", line 43, in _wrapper
    return bound_method(*args, **kwargs)
  File "/home/emes/venvs/reef/lib/python3.10/site-packages/django/utils/decorators.py", line 130, in _wrapped_view
    response = view_func(request, *args, **kwargs)
  File "/home/emes/venvs/reef/lib/python3.10/site-packages/constance/admin.py", line 256, in changelist_view
    form.save()
  File "/home/emes/venvs/reef/lib/python3.10/site-packages/constance/admin.py", line 157, in save
    setattr(config, name, new)
  File "/home/emes/venvs/reef/lib/python3.10/site-packages/django/utils/functional.py", line 277, in __setattr__
    setattr(self._wrapped, name, value)
  File "/home/emes/venvs/reef/lib/python3.10/site-packages/constance/base.py", line 29, in __setattr__
    self._backend.set(key, value)
  File "/home/emes/venvs/reef/lib/python3.10/site-packages/constance/backends/redisd.py", line 50, in set
    signals.config_updated.send(
  File "/home/emes/venvs/reef/lib/python3.10/site-packages/django/dispatch/dispatcher.py", line 180, in send
    return [
  File "/home/emes/venvs/reef/lib/python3.10/site-packages/django/dispatch/dispatcher.py", line 181, in <listcomp>
    (receiver, receiver(signal=self, sender=sender, **named))
  File "/home/emes/devel/reef/cct/project/app/src/project/core/constance_prometheus.py", line 134, in on_constance_config_updated
    store_monitored_metric(key, new_value)
  File "/home/emes/devel/reef/cct/project/app/src/project/core/constance_prometheus.py", line 117, in store_monitored_metric
    METRICS[key].store(new_value)
  File "/home/emes/devel/reef/cct/project/app/src/project/core/constance_prometheus.py", line 106, in _store_str
    self._metric = Gauge(self.name, self.description)
  File "/home/emes/venvs/reef/lib/python3.10/site-packages/prometheus_client/metrics.py", line 365, in __init__
    super().__init__(
  File "/home/emes/venvs/reef/lib/python3.10/site-packages/prometheus_client/metrics.py", line 143, in __init__
    registry.register(self)
  File "/home/emes/venvs/reef/lib/python3.10/site-packages/prometheus_client/registry.py", line 43, in register
    raise ValueError(

Exception Type: ValueError at /admin/constance/config/
Exception Value: Duplicated timeseries in CollectorRegistry: {'constance_config_SOME_STRING_hello_string'}
```

I wonder if we shouldn't drop the `str` (and other non-standard type) monitoring until we figure out a way to handle it better.

Reply:
Nevermind the above comment. I'll address the `"livemax"` mode later on.
BTW, have you noticed that `"livemax"` is available only in unreleased code? Latest official release (0.14.1) doesn't have it.